The excessive_precision linter suggestion causes a compilation error

### Summary

The lint suggests truncating `3.0000000000000000e+00` to `3` which causes a "mismatched types" compilation error.

### Lint Name

clippy::excessive_precision

### Reproducer

I tried this code:

```rust
pub const DATA: f64 = 3.0000000000000000e+00;
```

I saw this happen:

```
cargo clippy --fix --allow-staged
    Checking clippy_test v0.1.0 (C:\dev\tencek\sandbox\rust\clippy_test)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_test`

after fixes were automatically applied the compiler reported errors within these files:

  * src\data.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src\data.rs:1:23
  |
1 | pub const DATA: f64 = 3;
  |                       ^
  |                       |
  |                       expected `f64`, found integer
  |                       help: use a float literal: `3.0`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: float has excessive precision
 --> src\data.rs:1:23
  |
1 | pub const DATA: f64 = 3.0000000000000000e+00;
  |                       ^^^^^^^^^^^^^^^^^^^^^^ help: consider changing the type or truncating it to: `3`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#excessive_precision
  = note: `#[warn(clippy::excessive_precision)]` on by default

warning: `clippy_test` (bin "clippy_test") generated 1 warning (run `cargo clippy --fix --bin "clippy_test"` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_test`

after fixes were automatically applied the compiler reported errors within these files:

  * src\data.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src\data.rs:1:23
  |
1 | pub const DATA: f64 = 3;
  |                       ^
  |                       |
  |                       expected `f64`, found integer
  |                       help: use a float literal: `3.0`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.
```

I expected to see this happen:
Probably this:
```rust
pub const DATA: f64 = 3.0;
```

### Version

```text
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: x86_64-pc-windows-msvc
release: 1.79.0
LLVM version: 18.1.7
```


### Additional Labels

I-suggestion-causing-error

--- a/clippy_utils/src/numeric_literal.rs
+++ b/clippy_utils/src/numeric_literal.rs
@@ -164,6 +164,8 @@ impl<'a> NumericLiteral<'a> {
             if !exponent.is_empty() && exponent != "0" {
                 output.push_str(separator);
                 Self::group_digits(&mut output, exponent, group_size, true, false);
+            } else if exponent == "0" && self.fraction.is_none() && self.suffix.is_none() {
+                output.push_str(".0");
             }
         }
 

--- a/tests/ui/excessive_precision.fixed
+++ b/tests/ui/excessive_precision.fixed
@@ -78,4 +78,8 @@ fn main() {
     const NEG_INF1: f32 = -1.0e+33f32;
     const NEG_INF2: f64 = -1.0e+3300f64;
     const NEG_INF3: f32 = -3.40282357e+38_f32;
+
+    // issue #12954
+    const _: f64 = 3.0;
+    const _: f64 = 3.0000000000000000;
 }

--- a/tests/ui/excessive_precision.rs
+++ b/tests/ui/excessive_precision.rs
@@ -78,4 +78,8 @@ fn main() {
     const NEG_INF1: f32 = -1.0e+33f32;
     const NEG_INF2: f64 = -1.0e+3300f64;
     const NEG_INF3: f32 = -3.40282357e+38_f32;
+
+    // issue #12954
+    const _: f64 = 3.0000000000000000e+00;
+    const _: f64 = 3.0000000000000000;
 }

--- a/tests/ui/excessive_precision.stderr
+++ b/tests/ui/excessive_precision.stderr
@@ -91,5 +91,11 @@ error: float has excessive precision
 LL |     let _ = 1.000_000_000_000_001e-324_f64;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider changing the type or truncating it to: `0_f64`
 
-error: aborting due to 15 previous errors
+error: float has excessive precision
+  --> tests/ui/excessive_precision.rs:83:20
+   |
+LL |     const _: f64 = 3.0000000000000000e+00;
+   |                    ^^^^^^^^^^^^^^^^^^^^^^ help: consider changing the type or truncating it to: `3.0`
+
+error: aborting due to 16 previous errors
 

